memcache/memcache.go

```diff
@@ -313,14 +348,52 @@ func (c *Client) dial(addr net.Addr) (net.Conn, error) {
 	return nil, err
 }
 
+var timers = sync.Pool{
+	New: func() interface{} {
+		t := time.NewTimer(time.Hour)
+		t.Stop()
+		return t
+	},
+}
+
 func (c *Client) getConn(addr net.Addr) (*conn, error) {
+	if c.activeConnLock == nil {
+		c.lk.Lock()
+		if c.activeConnLock == nil {
+			c.activeConnLock = make(map[string]chan struct{})
+		}
+		c.lk.Unlock()
+	}
+	if c.activeConnLock[addr.String()] == nil {
+		c.lk.Lock()
+		if c.activeConnLock[addr.String()] == nil {
+			c.activeConnLock[addr.String()] = make(chan struct{}, c.maxConns())
+		}
+		c.lk.Unlock()
+	}
```

Comment:
I'd take this entire codeblock and stick it in a method off of `*Client` just for readability.